torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
   return v->cloneWithNewBody(body);
 }
 
+std::vector<Stmt*> GPUMetaVarRewriter::handleDependences(const Block* v) {
+  analysis::MemDependencyChecker analyzer;
+  v->accept(&analyzer);
+
+  // Iterate over the statements in the block. If any statement is dependent on
+  // any previous statement in the block, then insert a SyncThreads before that
+  // statement. Note that the iteration below is quadratic in the number of
+  // statements in the block.
```

Comment:
I do worry a little about adding an n^2 step here, since it operates on all statements - but I'm not 100% sure how else to do it.

Reply:
I don't have an alternative to this as of now. Added a TODO here. Will try to figure something out.